app/controllers/supplementary/DocumentsProducedController.scala

```diff
@@ -34,29 +41,130 @@ class DocumentsProducedController @Inject()(
-    form
-      .bindFromRequest()
-      .fold(
-        (formWithErrors: Form[DocumentsProduced]) =>
-          Future.successful(BadRequest(documents_produced(appConfig, formWithErrors))),
-        form =>
-          customsCacheService.cache[DocumentsProduced](supplementaryCacheId, formId, form).map { _ =>
-            Redirect(controllers.supplementary.routes.SummaryPageController.displayPage())
+    val boundForm = form.bindFromRequest()
+    val actionTypeOpt = request.body.asFormUrlEncoded.flatMap(FormAction.fromUrlEncoded(_))
+    val cachedData = customsCacheService
+      .fetchAndGetEntry[DocumentsProducedData](supplementaryCacheId, formId)
+      .map(_.getOrElse(DocumentsProducedData(Seq())))
+
+    cachedData.flatMap { cache =>
+      boundForm
+        .fold(
+          (formWithErrors: Form[DocumentsProduced]) =>
+            Future.successful(BadRequest(documents_produced(appConfig, formWithErrors, cache.documents))),
+          validForm =>
+            actionTypeOpt match {
+              case Some(Add)             => addItem(validForm, cache)
+              case Some(SaveAndContinue) => saveAndContinue(validForm, cache)
+              case Some(Remove(values))  => removeItem(retrieveItem(values), cache)
+              case _                     => errorHandler.displayErrorPage()
+          }
+        )
+    }
+  }
+
+  private def saveAndContinue(
+    userInput: DocumentsProduced,
+    cachedData: DocumentsProducedData
+  )(implicit request: AuthenticatedRequest[_], hc: HeaderCarrier): Future[Result] =
+    (userInput, cachedData.documents) match {
+      case (document, Seq())     => handleSaveAndContinueEmptyCache(document)
+      case (document, documents) => handleSaveAndContinueCache(document, documents)
+    }
+
+  private def handleSaveAndContinueCache(document: DocumentsProduced, documents: Seq[DocumentsProduced])(
+    implicit request: AuthenticatedRequest[_]
+  ) =
+    document match {
+      case _ if documents.length >= maxNumberOfItems =>
+        handleErrorPage(Seq(("", "supplementary.addDocument.maximumAmount.error")), document, documents)
+
+      case _ if documents.contains(document) =>
+        handleErrorPage(Seq(("", "supplementary.addDocument.duplicated")), document, documents)
+
+      case _ if document.isDefined =>
+        val updatedCache = DocumentsProducedData(documents :+ document)
+
+        customsCacheService.cache[DocumentsProducedData](supplementaryCacheId, formId, updatedCache).map { _ =>
+          Redirect(controllers.supplementary.routes.SummaryPageController.displayPage())
+        }
+
+      case _ if !document.isDefined =>
+        Future.successful(Redirect(controllers.supplementary.routes.SummaryPageController.displayPage()))
+    }
+
+  private def handleSaveAndContinueEmptyCache(document: DocumentsProduced)(implicit request: AuthenticatedRequest[_]) =
+    document match {
+      case _ if document.isDefined =>
```

Comment:
The same like above, here when the page is optional you don't need to use pattern matching, always will be redirected. Add only cache data when document is defined

Reply:
The golden rule for me don't use pattern matching for Booleans

Reply:
I am all about golden rules!!! Much better than silver, although copper is becoming quite fashionable these days.

Reply:
I have amended this. Please review and resolve both conversations if you are happy ;)

Reply:
I have addressed your concern about the `Future` and the `redirect` problem. Well spotted!